### PR TITLE
HGCAL validation at SimCluster level

### PR DESCRIPTION
**Pull request by lecriste**
#### PR description:

This PR introduces an HGCAL validation at SimCluster level. No changes are expected in the output apart from new DQM histograms.

#### PR validation:

`runTheMatrix -l limited`

### CODE REVIEW

**Comment by jfernan2** [comment]
please test

**Comment by kpedro88** [comment]
you need something like this to handle the premixing case: https://github.com/cms-sw/cmssw/blob/08fd7d924f8ef28001c9d0fec6df076fb368dce5/SimDataFormats/Associations/python/LCToCPAsssociation_cfi.py#L10-L12

**Comment by civanch** [comment]
please test

**Comment by srimanob** [comment]
@lecriste 
Have you tried to run it locally first before pushing the PR? It seems if you can run locally on a specific Phase-2 workflow, you should see the problem locally, i.e. 23434.999. 

Thanks.

**Comment by srimanob** [comment]
Please test

**Comment by apsallid** [comment]
I am trying to check this PR against the more recent RelVal closeby samples (as I always do). However, today running on CMSSW_11_3_X_2021-01-13-2300 jobs were failing with error [1]. This is something I saw for the first time and I suspected that from now on integration builds will need GPU nodes. However, opening a ticket in the 
service desk [2] I got a reply saying that GPUs are limited and CPU should be used instead. 

Is it true that running condor jobs on recent integration builds needs GPU nodes (so request_GPUs = 1 in the submit script) or am I missing something and error below could be overcomed in a CPU node also? 

Sorry for writing here, I don't know if i should ask in the hypernews computing tools instead.  

[1] 
---- Begin Fatal Exception 14-Jan-2021 12:16:45 CET-----------------------
An exception of category 'PluginLibraryLoadError' occurred while
   [0] Constructing the EventProcessor
   [1] While attempting to load plugin CaloTPGTranscoderULUTs
Exception Message:
unable to load /cvmfs/cms-ib.cern.ch/nweek-02663/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_3_X_2021-01-13-1100/lib/slc7_amd64_gcc900/pluginCalibCalorimetryCaloTPGPlugins.so because libcuda.so.1: cannot open shared object file: No such file or directory
----- End Fatal Exception -------------------------------------------------

[2] https://cern.service-now.com/service-portal?id=ticket&table=incident&n=INC2667627

**Comment by makortel** [comment]
> Is it true that running condor jobs on recent integration builds needs GPU nodes (so request_GPUs = 1 in the submit script) or am I missing something and error below could be overcomed in a CPU node also?

The way CUDA is currently set up in CMSSW should work on CPU-only nodes as well as long as no CUDA calls are being made. 

Can you check what `ldd /cvmfs/cms-ib.cern.ch/nweek-02663/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_3_X_2021-01-13-1100/lib/slc7_amd64_gcc900/pluginCalibCalorimetryCaloTPGPlugins.so | fgrep libcuda` shows in a setup that results in the exception?

@fwyzard Would you have any other suggestions, or thoughts what could be going wrong?

**Comment by apsallid** [comment]
> Can you check what `ldd /cvmfs/cms-ib.cern.ch/nweek-02663/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_3_X_2021-01-13-1100/lib/slc7_amd64_gcc900/pluginCalibCalorimetryCaloTPGPlugins.so | fgrep libcuda` shows in a setup that results in the exception?

This is really weird. I get 

	`libcudart.so.11.0` => /cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc900/cms/cmssw-patch/CMSSW_11_3_X_2021-01-13-2300/external/slc7_amd64_gcc900/lib/libcudart.so.11.0 (0x00007f9f85950000)
	libcuda.so.1 => /lib64/libcuda.so.1 `(0x00007f9f83aea000)`

So, the library that the error message complains that it cannot found, it is there. I am running the jobs with gpu nodes now normally, but they are processed really slow, 15 at time.

**Comment by fwyzard** [comment]
I'd say something is not behaving as expected.

On a machine without any GPUs, I get
```
$ ldd /cvmfs/cms-ib.cern.ch/nweek-02663/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_3_X_2021-01-13-1100/lib/slc7_amd64_gcc900/p
luginCalibCalorimetryCaloTPGPlugins.so | grep cuda 
        libcudart.so.11.0 => /cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc900/cms/cmssw-patch/CMSSW_11_3_X_2021-01-13-2300/external/slc7_amd64_gcc900/lib/libcudart.so.11.0 (0x00007fede335e000)
        libcuda.so.1 => /cvmfs/cms-ib.cern.ch/nweek-02663/slc7_amd64_gcc900/external/cuda/11.2.0-affcae8bc4fc6169fc68a741efad30d6/drivers/libcuda.so.1 (0x00007fede1362000)
```
so scram seems to be setting up "our" version of cuda correctly.

Is it possible that the node where you see this problem **has** a central CUDA installation, but one which is not up to date ?

**Comment by makortel** [comment]
Thanks, interesting. On a CPU-only node without any CUDA installation the `libcuda.so.1` dependence should resolve to `/cvmfs/cms-ib.cern.ch/nweek-02663/slc7_amd64_gcc900/external/cuda/11.2.0-affcae8bc4fc6169fc68a741efad30d6/drivers/libcuda.so.1`.

I wonder why that CPU-only node has CUDA drivers installed locally. Could the driver version be incompatible with CMSSW's CUDA runtime? Could you check `ls -l /lib64/libcuda.so.1`?

**Comment by fwyzard** [comment]
Also, if there is a `/lib64/libcuda.so.1`, you should not be getting the error `libcuda.so.1: cannot open shared object file: No such file or directory` ?

**Comment by apsallid** [comment]
> Could you check `ls -l /lib64/libcuda.so.1`?

In the lxplus machine I am connecting I get 

`lrwxrwxrwx. 1 root root 20 Jan  5 10:03 /lib64/libcuda.so.1 -> libcuda.so.450.80.02`

but you mean sent a job in a cpu only node and see that this prints out?

> Also, if there is a `/lib64/libcuda.so.1`, you should not be getting the error `libcuda.so.1: cannot open shared object file: No such file or directory` ?

Yes, this is really weird.

**Comment by apsallid** [comment]
@fwyzard @makortel Thanks a lot for your replies. It made me understand that we have an issue in our official  [HGCal production repository](https://github.com/CMS-HGCAL/reco-prodtools). The problem is in this [line](https://github.com/CMS-HGCAL/reco-prodtools/blob/master/SubmitFileGSD.sh#L22) where the `/afs/cern.ch/cms/cmsset_default.sh` is run. Commenting out the line runs fine without requesting gpu node.

**Comment by fwyzard** [comment]
Out of curiosity, what happens if instead of removing it, you change it to
```
source /cvmfs/cms.cern.ch/cmsset_default.sh
```
?

**Comment by makortel** [comment]
I noticed that (all?) lxplus nodes indeed have `/lib64/libcuda.so.450.80.02`. But
```
$ cmsrel CMSSW_11_3_X_2021-01-13-2300
$ cd CMSSW_11_3_X_2021-01-13-2300/src
$ cmsenv
$ ldd /cvmfs/cms-ib.cern.ch/nweek-02663/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_3_X_2021-01-13-1100/lib/slc7_amd64_gcc900/pluginCalibCalorimetryCaloTPGPlugins.so | fgrep cuda
        libcudart.so.11.0 => /cvmfs/cms-ib.cern.ch/week1/slc7_amd64_gcc900/cms/cmssw-patch/CMSSW_11_3_X_2021-01-13-2300/external/slc7_amd64_gcc900/lib/libcudart.so.11.0 (0x00007f9835c3c000)
        libcuda.so.1 => /cvmfs/cms-ib.cern.ch/nweek-02663/slc7_amd64_gcc900/external/cuda/11.2.0-affcae8bc4fc6169fc68a741efad30d6/drivers/libcuda.so.1 (0x00007f9833c40000)
```

If I prepend `/lib64` to `LD_LIBRARY_PATH` after `cmsenv` I can reproduce `ldd ... | fgrep cuda` showing `/lib64/libcuda.so.1` (but then it also shows lots of warnings from incompatible glibc etc so the conditions leading to the exception are not that simple).

**Comment by davidlange6** [comment]
Yea /afs/cern.ch/cms/cmsset_default.sh misses a few years worth of updates by now.

**Comment by fwyzard** [comment]
In particular, it's missing an update where we told scram to use the bundled version of `libcuda.so` if the system one was too old.

**Comment by davidlange6** [comment]
@smuzaffar - can we just remove /afs/cern.ch/cms/cmsset_default.sh?

**Comment by apsallid** [comment]
> Out of curiosity, what happens if instead of removing it, you change it to
> 
> ```
> source /cvmfs/cms.cern.ch/cmsset_default.sh
> ```
> 
> ?

It runs just fine, no problem at all.

**Comment by smuzaffar** [comment]
@apsallid , the dynamic setting of cuda env feature is supported by new versions of scram (which are only available on cvmfs). Once you source `/afs/cern.ch/cms/cmsset_default.sh` then basically you are using an very old version of scram which does not support this feature. 

@davidlange6 , getting rid of `/afs/cern.ch/cms/cmsset_default.sh` sounds good. I think as a first step we just rename these scripts on afs and after few week are can delete

**Comment by lecriste** [comment]
@civanch can you please review this PR?

**Comment by lecriste** [comment]
@cms-sw/simulation-l2 can you please review this PR?

File: Validation/HGCalValidation/scripts/makeHGCalValidationPlots.py [hunk 1]
```diff
@@ -48,14 +51,23 @@ def main(opts):
         val = SeparateValidation([sample], opts.outputDir[0])
     htmlReport = val.createHtmlReport(validationName=opts.html_validation_name[0])   
 
+    #layerClusters
     if opts.collection==layerClustersGeneralLabel:
 	hgclayclus = [hgcalPlots.hgcalLayerClustersPlotter]
 	hgcalPlots.append_hgcalLayerClustersPlots("hgcalLayerClusters", "Layer Clusters", extendedFlag)
         val.doPlots(hgclayclus, plotterDrawArgs=drawArgs)
+    #simClusters
+    elif (opts.collection == simClustersGeneralLabel) :
```
Comment (srimanob) [inline Validation/HGCalValidation/scripts/makeHGCalValidationPlots.py:60]: Not necessary, but can we try to make coding in the same style, at least in the same file? For example, we have a mixing between
- opts.collection==layerClustersGeneralLabel:
- (opts.collection == simClustersGeneralLabel) :

Comment (lecriste) [reply Validation/HGCalValidation/scripts/makeHGCalValidationPlots.py]: https://github.com/cms-sw/cmssw/commit/34d750f072c38954edf9eb0699ac505517583778

**Comment by srimanob** [comment]
Please test

**Comment by jfernan2** [comment]
@lecriste  this PR is adding ~15k new MEs for HGCAL accounting for 70MB of extra memory, on top of the already existing ~8k MEs using 62MB for HGCAL, so doubling the current load for HGCAL
The memory and CPU budget for DQM is already quite low, would there be any chance to reduce the additions of this PR to those absolutely needed?

**Comment by lecriste** [comment]
@jfernan2 [these](https://github.com/cms-sw/cmssw/blob/34d750f072c38954edf9eb0699ac505517583778/Validation/HGCalValidation/interface/HGVHistoProducerAlgo.h#L118-L139) are the core MEs added and I don't think they can be dropped without loosing important information.
The "problem" is the layer multiplicity for each of those vector, and the number of [iterations](https://github.com/cms-sw/cmssw/blob/ef3b75f6bef863a77fa0ae7aad789a3d482951af/Validation/HGCalValidation/python/HGCalValidator_cfi.py#L53-L59) for which they are produced; none of them can be reduced.

**Comment by jfernan2** [comment]
Would it be possible to at least reduce the binning in order to decrease the memory footprint?

**Comment by lecriste** [comment]
@jfernan2 the binning of the simCluster MEs introduced by this PR is in common with other MEs (layerCluster, Multicluster, ecc.), and is adapted to the HGCAL granularity.
However I found that the phi MEs range was 20% wider than needed, the last commit should have a sizeable effect on the HGCAL memory footprint.

**Comment by srimanob** [comment]
Please test

**Comment by jfernan2** [comment]
> However I found that the phi MEs range was 20% wider than needed, the last commit should have a sizeable effect on the HGCAL memory footprint.


Thanks @lecriste but the change only made a small reduction (500KB) in memory footprint:
Before: DQMHistoSizes: changed ( 23234.0,... ): 69597.749 KiB HGCAL/HGCalValidator
Now: DQMHistoSizes: changed ( 23234.0,... ): 69202.827 KiB HGCAL/HGCalValidator

**Comment by lecriste** [comment]
@jfernan2 I found another variable whose binning was out of range, and pushed a fix commit.
Hope this time the memory reduction is larger, since it affects some 2D ME.
@cmsbuild please test

**Comment by jfernan2** [comment]
please test

**Comment by jfernan2** [comment]
Thanks @lecriste

**Comment by lecriste** [comment]
@jfernan2 the last commit reduces the HGCAL memory footprint sizeably.

**Comment by jfernan2** [comment]
@lecriste  there are several MEs with no entries in the workflows tested, is this expected due to the reduced stats?
https://tinyurl.com/yydog8gm
https://tinyurl.com/y6zt2ll3
https://tinyurl.com/y4vl6klt
and MEs around in that and other folders

**Comment by lecriste** [comment]
With reduced stas only MEs of the first layers are filled, and for not all HGCAL iterations (ticlTracksters*).
A test with more events can be found [here](https://apsallid.web.cern.ch/apsallid/associatorsimClusterPlots/HGCValid_simClusters_Plots_CE_E_Front_200um).

Regarding the `score` MEs, they are expected to be empty as we did not implement a score computation yet. `Duplicates` histograms are empty but `efficiency` are filled, same for `fake` and `merged`.
We would like to have this PR merged and have a more careful design of the MEs later.

**Comment by jfernan2** [comment]
+1

**Comment by lecriste** [comment]
@cms-sw/simulation-l2 can you please review this PR?

**Comment by lecriste** [comment]
@civanch can you please review this PR?

**Comment by srimanob** [comment]
+Upgrade

If there is no concern about empty things flying around from DQM, I think it should be OK. Additional PRs to come after a careful design of MEs.

**Comment by civanch** [comment]
+1

**Comment by silviodonato** [comment]
+1

**Pull request merged**